.github/workflows/build.yml

```
@@ -0,0 +1,64 @@
+name: Build struct2tensor wheels
+
+on:
+  pull_request:
+  workflow_dispatch:
+  release:
+    types: [published]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Build package
+        id: build-package
+        uses: ./.github/reusable-build
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: List and check wheels
+        shell: bash
+        run: |
+          # Pin transitive dependency on pkginfo until https://github.com/pypa/twine/issues/1070 is
+          # fixed
+          pip install twine pkginfo>=1.10.0
+          ${{ matrix.ls || 'ls -lh' }} dist/
+          twine check dist/*
+
+  upload_to_pypi:
+    name: Upload to PyPI
+    runs-on: ubuntu-latest
+    if: (github.event_name == 'release' && startsWith(github.ref, 'refs/tags')) || (github.event_name == 'workflow_dispatch')
```

```suggestion
    if: github.event_name == 'release' && github.event.action == 'published') || (github.event_name == 'workflow_dispatch')
```

A GitHub release will always come from a tag, but a tag might not be associated with a GitHub release until someone creates one from it – this suggestion makes the event comply with what is mentioned in the PR description.

---

Also, side note: personally, I feel it's better not to allow the `workflow_dispatch:` event to trigger the PyPI upload job since those with "triage" access on GitHub and higher (instead of "maintain" access or higher) will be able to trigger it. It might be worth restricting this with at least one required reviewer (or two) in the environment settings on GitHub as an additional measure (I assume that has been done already, though).

---

This workflow only triggers on published releases currently, but this is safer by default - if someone else edits this action and adds additional release-related events we could trigger here, so I'm happy to make this change. And your point about `workflow_dispatch` is important - I'll get rid of the `workflow_dispatch` trigger right away!